stdlib/public/core/StringBridge.swift

```diff
@@ -795,19 +795,33 @@ extension StringProtocol {
   @_specialize(where Self == Substring)
```

Comment:
Given the dynamic checks, do we still need the `@_specialize` annotations here, since they're effectively doing the same thing?